### ISSUE TITLE
OCR0010: Test the TrOCR's performance (5)

### ISSUE THREAD

**Issue by ta4tsering** [OpenPecha/TrOCR-Pipeline#1]
Description:

- Evaluate the performance of TrOCR (Transformer-based OCR) by comparing the use of transliteration data versus Unicode data as training data.

The process involves:
- Data Preparation
- Training 2 separate TrOCR models using transliteration and Unicode data.
- The model for Unicode is already present.
- Defining evaluation metrics and Performance comparison.

Completion Criteria:

- TrOCR models trained using transliteration data and Unicode data.
- Comprehensive quantifiable report comparing TrOCR performance with transliteration data versus Unicode data.

Implementation plan:

![Image](https://github.com/OpenPecha/TrOCR-Pipeline/assets/43548581/a9c806b5-2c91-4be6-b7a7-f14b3ed1f36c)

subtask:
- [x] Reading and understanding the TrOCR paper [TrOCR: Transformer-based Optical Character Recognition
with Pre-trained Models](https://arxiv.org/pdf/2109.10282v5.pdf)
- [x] familiarizing and updating the existing script [TrOCR Pipeline](https://github.com/OpenPecha/TrOCR-Pipeline)
- [x] Creating the environment for the test on the vast ai
- [x]  Downloading one woodblock data ( Lhasa Kanjur)
- [x] playing with the hyperparameters
- [x] training TrOCR on the unicode text using [tibetan-roberta-base](https://huggingface.co/sangjeedondrub/tibetan-roberta-base)
- [ ] Generate the comparison between two using the benchmark or test set

**Comment by ta4tsering** [comment]
decided on using only the Lhasa Kanjur data for the testing so that it doesn't cost time as well as money after consulting with the team.

**Comment by ta4tsering** [comment]
writing script to split the data into its own folders as per the split provided by eric in the s3 bucket, need to update the TrOCR pipeline a little bit as it was written in mind that the split between the eval and train set happening on the go, not beforehand which is the case for now, so created a new branch and updating the script.

**Comment by ta4tsering** [comment]
I am having issues with the feature extractor that I am using for the 

![Image](https://github.com/OpenPecha/TrOCR-Pipeline/assets/43548581/59364dd0-309a-4401-b9bd-37af43f1b0bc)

**Comment by ta4tsering** [comment]
the issue was the the `max_target_length` for which I had to find the max token count in a sentence after tokenized, so for that I tokenized all the sentences in the train set and then take the highest number of tokens in a tokenized sentence, which was `170` and then updated the max_taget_length to 171. that fixed the issue and then I also use 
`Seq2SeqTrainingArguments(
    predict_with_generate=True,
    evaluation_strategy="steps",
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    fp16=True, 
    output_dir="TrOCR-Tibetan",
    logging_steps=100,
    save_steps=4000,
    eval_steps=4000,
    report_to=["wandb"]
)` for the arguments and then 

`trainer = Seq2SeqTrainer(
    model=model,
    tokenizer=processor.feature_extractor,
    args=training_args,
    compute_metrics=compute_metrics,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    data_collator=default_data_collator,
)` for trainer. which are Transformer API. Also adding `wandb` so that I can see the graphs as well.

**Comment by ta4tsering** [comment]
@spsither To test or review the https://github.com/OpenPecha/TrOCR-Pipeline/blob/main/Fine_tune_TrOCR_pipeline.ipynb notebook script and then script to upload the model and processor https://github.com/OpenPecha/TrOCR-Pipeline/blob/main/upload_model_and_processor.ipynb and then uploaded model and processor are at https://huggingface.co/ta4tsering/Lhasa_Kanjur_TrOCR_model-44000 and https://huggingface.co/ta4tsering/Lhasa_Kanjur_TrOCR_model

**Comment by spsither** [comment]
Use [push_to_hub](https://huggingface.co/docs/diffusers/en/using-diffusers/push_to_hub) in `upload_model_and_processor.ipynb`
